### PR TITLE
add recompute for attention and mhc_forward

### PR DESCRIPTION
**Pull request by LiYuRio**
### PR Category

Performance Optimization

### PR Types

Performance

### Description

增加attention的recompute

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py [hunk 1]
```diff
@@ -761,6 +768,165 @@ def forward(
                 dense_mode=self.config.csa_dense_mode,
             )
 
+        # Full attention recompute: wrap qkv + core_attn + inv_rope + o_group_proj + gated_attn
+        # into one RecomputeWithoutOutput block. All intermediates (query, key, value, etc.)
+        # are freed after forward (no_grad), and the output is discarded after o_proj.
+        if self.recompute_full_attn and self.training:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dsv4_hybrid_attention.py:776]: P1 `full_attn` 的评估分支绕过了 FP8 grouped output

这里加上 `self.training` 后，评估态会进入下方直接分支；该分支固定执行 `paddle.einsum`，而 `_full_attn_forward` 在 `full_fp8_computation && FLEET_FP8_WO_A_GEMM` 时会调用 `GroupedOutputFP8.apply`。因此同一配置会出现训练使用 FP8、评估静默退回普通 grouped projection 的行为差异，`full_fp8_computation` 在评估时不再生效。

请将 grouped-output projection 抽成两条路径共用的 helper（或在直接分支同步 FP8 条件），并补一个 `eval()` + `full_fp8_computation=True` 的分支回归测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [dsv4_hybrid_attention.py:776](https://github.com/PaddlePaddle/PaddleFleet/pull/1595#discussion_r3683271735) | P1 | ✅ |
| 2 | [dsv4_hybrid_attention.py:944](https://github.com/PaddlePaddle/PaddleFleet/pull/1595#discussion_r3683362445) | P1 | ✅ |
| 3 | [transformer_layer.py:1365](https://github.com/PaddlePaddle/PaddleFleet/pull/1595#discussion_r3683770125) | P1 | ✅ |
| 4 | [test_ai_hyper_connection_transformer_layer.py:975](https://github.com/PaddlePaddle/PaddleFleet/pull/1595#discussion_r3683771935) | P1 | ✅ |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py [hunk 2]
```diff
@@ -859,14 +931,18 @@ def forward(
...
             )
-            if self.recompute_gated_attn:
-                gate_recompute = RecomputeWithoutOutput()
-                core_attn_out = gate_recompute.recompute(
+            # When NOT inside full_attn recompute, gated_attn can have its own
+            # independent RecomputeWithoutOutput wrapper for lighter memory saving.
+            if (
+                self.recompute_gated_attn
+                and self.training
+                and self._full_attn_recompute is None
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dsv4_hybrid_attention.py]: P1 full-attention 的 backward 重算会误启用嵌套 gated 重算

外层 full-attention forward 在注册 hook 后会把 `self._full_attn_recompute` 置空；backward 时其 hook 会再次调用 `_full_attn_forward`，因此这里的条件变为真并新建 `_gate_recompute`。但这次调用不会返回到 `forward()` 的 809-817 行执行 `discard_output_and_register_recompute`，嵌套 PyLayer 的 `ctx.inputs/ctx.outputs` 无法由 hook 填充，`full_attn + gated_attn` 组合在反向传播时会失败。

请把“是否处于 full-attention 重算”作为显式参数传给 helper（forward 和 backward 重算保持同值），或只在顶层 `forward()` 包 gated 重算；同时补一个同时启用两项并实际调用 `backward()`、校验梯度的测试。

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/dsv4_hybrid_attention.py:776]: 当前提交已让评估分支复用 `_full_attn_forward`，FP8 grouped projection 的条件不再丢失；此问题已修复。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py (hunk 2 above)
Comment (LiYuRio) [reply src/paddlefleet/transformer/dsv4_hybrid_attention.py]: done

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py (hunk 1 above)
Comment (LiYuRio) [reply src/paddlefleet/transformer/dsv4_hybrid_attention.py:776]: done

File: tests/single_card_tests/ai_edited_test/transformer/test_ai_dsv4_hybrid_attention_recompute.py [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+for _k in _stale:
+    del sys.modules[_k]
+
+import types as _types
+import unittest
+from unittest.mock import MagicMock, patch
+
+import paddle
+
+from paddlefleet.tensor_parallel import RecomputeWithoutOutput
```
[the remark this answers is not in the document]
Comment (LiYuRio) [reply tests/single_card_tests/ai_edited_test/transformer/test_ai_dsv4_hybrid_attention_recompute.py]: done

File: tests/single_card_tests/ai_edited_test/transformer/test_ai_dsv4_hybrid_attention_recompute.py [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+    )
+)
+_src_dir = os.path.join(_repo_root, "src")
+sys.path.insert(0, _src_dir)
+
+# Evict any pre-loaded paddlefleet modules so the local source is used
+_stale = [k for k in sys.modules if k.startswith("paddlefleet")]
+for _k in _stale:
+    del sys.modules[_k]
+
```
[the remark this answers is not in the document]
Comment (LiYuRio) [reply tests/single_card_tests/ai_edited_test/transformer/test_ai_dsv4_hybrid_attention_recompute.py]: done

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py [hunk 5]
```diff
@@ -701,6 +701,12 @@ def __init__(
             and "gated_attn" in config.recompute_modules
         )
 
+        self.recompute_full_attn = (
+            config.recompute_granularity == "selective"
+            and config.recompute_modules is not None
+            and "full_attn" in config.recompute_modules
+        )
+
```
[the remark this answers is not in the document]
Comment (LiYuRio) [reply src/paddlefleet/transformer/dsv4_hybrid_attention.py:711]: done

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/dsv4_hybrid_attention.py]: 当前提交通过显式传入 `_in_full_recompute=True` 修复了嵌套重算的实现问题，但回归测试仍未闭环：`test_gated_attn_not_triggered_when_full_attn_active` 只用 mock 执行 forward，没有调用真实 `RecomputeWithoutOutput` 的 backward 或校验梯度；并且新增第 6 个位置参数后，测试中的 `len(args) == 5` 会把 full-attention 调用计入 `gate_recompute_count`，当前断言会失败。请同步修正该测试，并补充 `full_attn + gated_attn` 的真实 backward/梯度回归后再关闭此项。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/transformer_layer.py [hunk 6]
```diff
@@ -1357,6 +1358,14 @@ def __init__(
         # idempotent: already-colored base params are skipped.
         self._mark_shared_no_hook_params()
 
+        # mHC forward recompute config
+        self.recompute_mhc_forward = (
+            config.recompute_granularity == "selective"
+            and config.recompute_modules is not None
+            and isinstance(config.recompute_modules, list)
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_layer.py]: P1 `mhc_forward` 未遵循按层 selective recompute 配置

`recompute_modules` 的公开契约同时支持 list 和 dict；现有 `norm`/`mlp` 分支也会结合 `recompute_num_layers`、`recompute_method` 判断当前层。这里仅检查 list membership，导致 list + `recompute_num_layers` 时所有 mHC 层都开启重算，而 `{"mhc_forward": N}` 会被静默关闭，实际显存策略与配置不符。

请复用 `need_recompute_in_block` / `need_recompute_in_first_n` 处理 list 的按层模式和 dict 模式，并分别补测未命中的层与命中的层。

File: tests/single_card_tests/ai_edited_test/transformer/test_ai_hyper_connection_transformer_layer.py [hunk 7]
```diff
@@ -919,6 +923,68 @@ def test_selective_recompute_norm_forward(self):
...
+        )
+        layer = _make_hc_layer(config)
+        layer.train()
+        B, S = 2, 4
+        n = config.num_residual_streams
+        C = config.hidden_size
+        hidden_states = paddle.randn([S, B, n * C])
+        hidden_states.stop_gradient = False
+        dict_args = {"hidden_states": hidden_states, "attention_mask": None}
+        result = layer.forward(dict_args)
```
Comment (risemeup1111) [inline tests/single_card_tests/ai_edited_test/transformer/test_ai_hyper_connection_transformer_layer.py:971]: P1 该测试没有触发 mHC 重算的 backward hook

实现会清空 `aggregated`、`h_res`、`h_post`，其正确性依赖输出 hook 在反向阶段重建这三个张量；但这里仅执行 forward 并检查 shape/NaN/成员清理，无法发现多输出重算或 attention→MLP 两段 hook 链的梯度错误。

请使用真实 `RecomputeWithoutOutput` 调用 `loss.backward()`，并用相同权重和输入对比开启/关闭 `mhc_forward` 时的输出、输入梯度及 mHC 参数梯度。

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/dsv4_hybrid_attention.py]: 本次提交修正了 mock 对新增位置参数的计数，但仍未增加 `full_attn + gated_attn` 使用真实 `RecomputeWithoutOutput` 的 backward/梯度测试；原问题仍是部分修复，继续保留。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/transformer_layer.py (hunk 6 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_layer.py]: 当前提交已复用 `need_recompute_in_block` / `need_recompute_in_first_n` 处理 list 的按层模式和 dict 模式，并补充了命中与未命中层的测试；此项已修复。

File: tests/single_card_tests/ai_edited_test/transformer/test_ai_hyper_connection_transformer_layer.py (hunk 7 above)
Comment (risemeup1111) [reply tests/single_card_tests/ai_edited_test/transformer/test_ai_hyper_connection_transformer_layer.py:971]: 当前提交已实际执行 backward 并比较输入梯度，但第 1079 行仅在两侧 `grad` 均非 `None` 时比较；如果重算侧丢失 mHC 参数梯度，测试会直接跳过，仍不能验证原要求的参数梯度。请针对两个 hyper-connection 模块的 `mapping_proj.weight`、`alpha_*` 和 `bias`，先逐项断言两侧梯度非 `None`，再执行 `allclose`。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/dsv4_hybrid_attention.py]: 本次提交只调整了 DSv4 测试的导入路径；该文件仍没有调用 `.backward()`、`paddle.grad` 或 `autograd.backward`，现有重算测试也仍使用 forward-only mock。`full_attn + gated_attn` 的真实 backward/梯度回归尚未补充，此项继续保留。

File: tests/single_card_tests/ai_edited_test/transformer/test_ai_hyper_connection_transformer_layer.py (hunk 7 above)
Comment (risemeup1111) [reply tests/single_card_tests/ai_edited_test/transformer/test_ai_hyper_connection_transformer_layer.py:971]: 当前提交已逐项覆盖两个 hyper-connection 模块的 `mapping_proj.weight`、`alpha_*` 和 `bias`，先断言重算与非重算两侧梯度均非 `None`，再比较梯度一致性；此项已修复。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by Difers — DISMISSED** [review]

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/dsv4_hybrid_attention.py]: 当前提交已补充 `full_attn + gated_attn` 的真实 `RecomputeWithoutOutput` 回归：训练态同时启用两项，执行 forward/backward，并与无重算路径比较输出、输入梯度及 gate、grouped output、Q/KV 投影等关键参数梯度；此项已修复。

**Review by risemeup1111 — APPROVED** [review]

**Review by zjjlivein — APPROVED** [review]

**Comment by risemeup1111** [comment]
✅ Cherry-pick successful! Created PR: https://github.com/PaddlePaddle/PaddleFleet/pull/1610
